### PR TITLE
Asset Conversion pallet

### PR DESCRIPTION
**Pull request by gilescope**
A simple automated market maker pallet.

I'd direct both non-technical discussion and discussion of possible future improvements to be hosted on the forum: https://forum.polkadot.network/t/statemint-e-swap-discussion/1498

The goal of this pallet is to provide people with a simple audited AMM pallet should a substrate chain need a basic decentralised exchange.

### CODE REVIEW

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2180709

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2180708

File: primitives/rpc/Cargo.toml [hunk 1]
```diff
@@ -1,6 +1,6 @@
 [package]
 name = "sp-rpc"
-version = "6.0.0"
+version = "7.0.0"
```
Comment (bkchr) [inline primitives/rpc/Cargo.toml:3]: Why?

File: bin/node/executor/Cargo.toml [hunk 2]
```diff
@@ -39,6 +39,7 @@ pallet-sudo = { version = "4.0.0-dev", path = "../../../frame/sudo" }
 pallet-timestamp = { version = "4.0.0-dev", path = "../../../frame/timestamp" }
 pallet-treasury = { version = "4.0.0-dev", path = "../../../frame/treasury" }
 pallet-transaction-payment = { version = "4.0.0-dev", path = "../../../frame/transaction-payment" }
+pallet-dex = { version = "4.0.0-dev", path = "../../../frame/dex" }
```
Comment (bkchr) [inline bin/node/executor/Cargo.toml:42]: Same question?

File: frame/dex/rpc/runtime-api/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,31 @@
+// This file is part of Substrate.
```
Comment (xlc) [inline frame/dex/rpc/runtime-api/src/lib.rs:1]: didn't we deprecated custom RPC?

File: frame/dex/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,855 @@
...
+
+	#[pallet::pallet]
+	pub struct Pallet<T>(_);
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+
+		/// Units are 10ths of a percent
+		type Fee: Get<u64>;
```
Comment (xlc) [inline frame/dex/src/lib.rs:80]: should be `Permill`

File: frame/dex/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,855 @@
...
+		pub fn add_liquidity(
+			origin: OriginFor<T>,
+			asset1: MultiAssetId<T::AssetId>,
+			asset2: MultiAssetId<T::AssetId>,
+			amount1_desired: AssetBalanceOf<T>,
+			amount2_desired: AssetBalanceOf<T>,
+			amount1_min: AssetBalanceOf<T>,
+			amount2_min: AssetBalanceOf<T>,
+			mint_to: T::AccountId,
+			deadline: T::BlockNumber,
```
Comment (xlc) [inline frame/dex/src/lib.rs:306]: there is no need to have deadline as we already have transaction mortality

File: frame/dex/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,855 @@
...
+			}
+			Ok(())
+		}
+
+		/// The account ID of the pool.
+		///
+		/// This actually does computation. If you need to keep using it, then make sure you cache
+		/// the value and only call this once.
+		pub fn get_pool_account(pool_id: PoolIdOf<T>) -> T::AccountId {
+			let sub = T::Hashing::hash_of(&(T::PalletId::get(), pool_id)).encode();
```
Comment (xlc) [inline frame/dex/src/lib.rs:611]: why reimplement into_sub_account

File: frame/dex/src/lib.rs [hunk 7]
```diff
@@ -0,0 +1,855 @@
...
+		fn mul_div(
+			a: &AssetBalanceOf<T>,
+			b: &AssetBalanceOf<T>,
+			c: &AssetBalanceOf<T>,
+		) -> Result<AssetBalanceOf<T>, Error<T>> {
+			let a = u128::try_from(*a).map_err(|_| Error::<T>::UnsupportedBalanceType)?;
+			let b = u128::try_from(*b).map_err(|_| Error::<T>::UnsupportedBalanceType)?;
+			let c = u128::try_from(*c).map_err(|_| Error::<T>::UnsupportedBalanceType)?;
+
+			let result = a
```
Comment (xlc) [inline frame/dex/src/lib.rs:762]: this need to promote to U256 to avoid overflow, which is going to happen

**Comment by xlc** [comment]
There is no good reason to support Currency & Assets, which reduces compatibility and add complexity. If you really want to implement DEX, I will suggest take a look to some of the existing implementations first.

File: primitives/rpc/Cargo.toml (hunk 1 above)
Comment (gilescope) [reply primitives/rpc/Cargo.toml:3]: `sc-tracing` is using sp-rpc v7.

File: frame/dex/src/lib.rs (hunk 7 above)
Comment (gilescope) [reply frame/dex/src/lib.rs:762]: It's promoting from u64 already. U256 might be overkill, but types are plugable so maybe we need a trait to be able to promote into a larger numeric type? (and a trait to try and get back)

Comment (xlc) [reply frame/dex/src/lib.rs:762]: you have to support u128 balance type

**Comment by gilescope** [comment]
It's no secret that the dex pallet might well be used in the statemint runtime. That runtime currently uses the Balances and Assets pallets. All the bounds are against traits so it shouldn't preclude the dex being backed by other asset registry pallets that support the `fungibles` traits.

File: frame/dex/src/lib.rs (hunk 5 above)
Comment (gilescope) [reply frame/dex/src/lib.rs]: Thanks - it makes sense that this functionality is generic and available to every transaction.

File: bin/node/runtime/src/lib.rs [hunk 8]
```diff
@@ -1484,6 +1485,45 @@ impl pallet_assets::Config for Runtime {
 	type BenchmarkHelper = ();
 }
 
+impl pallet_assets::Config<Instance2> for Runtime {
+	type RuntimeEvent = RuntimeEvent;
+	type Balance = u128;
+	type AssetId = u32;
+	type Currency = Balances;
+	type CreateOrigin = AsEnsureOriginWithArg<EnsureSigned<AccountId>>;
```
Comment (joepetrowski) [inline bin/node/runtime/src/lib.rs:1493]: This needs to be the DEX pallet, this instance is not for users to register assets.

File: frame/dex/rpc/runtime-api/src/lib.rs (hunk 3 above)
Comment (joepetrowski) [reply frame/dex/rpc/runtime-api/src/lib.rs:1]: Yeah, anything new should be behind `state_call`

File: bin/node/runtime/src/lib.rs [hunk 9]
```diff
@@ -1534,6 +1574,9 @@ impl pallet_uniques::Config for Runtime {
 	type CollectionId = u32;
 	type ItemId = u32;
 	type Currency = Balances;
+	type CurrencyBalance = <Self as pallet_balances::Config>::Balance;
+	type Assets = Assets;
+	type AssetId = <Self as pallet_assets::Config<Instance1>>::AssetId;
```
Comment (joepetrowski) [inline bin/node/runtime/src/lib.rs:1579]: Why the change to the `pallet_uniques` config?

File: frame/dex/src/lib.rs (hunk 6 above)
Comment (gilescope) [reply frame/dex/src/lib.rs:592]: `into_sub_account_truncating` when given a `pool_id` hash manages to truncate the hash such that the pool accounts collide. Using the safer `try_into_sub_account` method returns `None` for the hash. It seems those functions focus on the first 4 bytes and disregard the rest? It's essential that the pool accounts remain separate and don't collide - I'm open to suggestions.

File: bin/node/runtime/src/lib.rs (hunk 9 above)
Comment (gilescope) [reply bin/node/runtime/src/lib.rs:1579]: Sorry this should have been reverted. Child PR to fix that here: https://github.com/paritytech/substrate/pull/12991

Comment (gilescope) [reply bin/node/runtime/src/lib.rs]: Fixed

File: bin/node/runtime/src/lib.rs (hunk 8 above)
Comment (gilescope) [reply bin/node/runtime/src/lib.rs]: Agreed the kitchen sink should set a good example. Fixed.

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2184268

**Comment by xlc** [comment]
You need to check how orml-tokens, orml-currencies works. I did a workshop about it https://github.com/xlc/orml-workshop
Basically the DEX pallet shouldn't need to handle two currency implementation. It is an adapter's job to handle that.

File: frame/dex/src/lib.rs (hunk 6 above)
Comment (xlc) [reply frame/dex/src/lib.rs:592]: The problem is your mock config https://github.com/paritytech/substrate/pull/12984/files#diff-093e9ce3d40b1593d70b5d0e418ec01c572a206d16ff17d4ff266eb34d9356b9R66
Make it u128.

File: primitives/rpc/Cargo.toml (hunk 1 above)
Comment (bkchr) [reply primitives/rpc/Cargo.toml:3]: How does that work on master? Something seems to be wrong here or something maybe pulls the dep now from crates.io?

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2186041

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2186098

**Comment by gilescope** [comment]
@xlc I hear you. I'm hoping that XCM v3 which is "coming soon™" gets merged. When that happens I'd like to re-work the interface so that MultiAssetId is replaced with XCM v3's MultiLocation as that can represent everything in a standard way.

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2191615

File: primitives/rpc/Cargo.toml (hunk 1 above)
Comment (gilescope) [reply primitives/rpc/Cargo.toml]: As we don't need the RPC crate we don't need to bump this version as we're not pulling in `sc-tracing` any more so this resolves.

File: bin/node/executor/Cargo.toml (hunk 2 above)
Comment (gilescope) [reply bin/node/executor/Cargo.toml]: Not needed now, removed.

File: frame/dex/rpc/runtime-api/src/lib.rs (hunk 3 above)
Comment (gilescope) [reply frame/dex/rpc/runtime-api/src/lib.rs]: Thanks for pointing that out @xlc, have removed the RPC and tested out that `state_call` works as expected.

File: frame/dex/src/lib.rs (hunk 6 above)
Comment (ggwpez) [reply frame/dex/src/lib.rs:599]: Additionally you can add a test in the `integrity_test` hook so that this never gets executed configured with a runtime with too short account ids.

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2291953

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2292004

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2297194

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2298685

**Review by joepetrowski — COMMENTED** [review]
A lot of documentation needs improvement.

**Comment by Chralt98** [comment]
> @xlc I hear you. I'm hoping that XCM v3 which is "coming soon™" gets merged. When that happens I'd like to re-work the interface so that MultiAssetId is replaced with XCM v3's MultiLocation as that can represent everything in a standard way.

So we can expect that everything works with orml_tokens and orml_currencies in the first iteration of this pallet? This would be super useful, especially because many projects depend on the orml library. pallet_asset support is fine, but the real fruition kicks in with MultiAsset pallets like from the orml library. pallet_assets differs from orml_tokens in key points as we know. I don't think that pallet_assets is meant to replace orml_tokens, right?

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2383777

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2383776

**Comment by jsidorenko** [comment]
bot rebase

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable-int
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2487666

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable-int
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2533392

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable-int
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2542083

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2542076

**Comment by jsidorenko** [comment]
bot clean

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable-int
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2551632

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable-int
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2571383

**Comment by jsidorenko** [comment]
bot clean

File: frame/dex/src/lib.rs [hunk 10]
```diff
@@ -0,0 +1,849 @@
...
+		/// Returns a pool id constructed from 2 sorted assets.
+		pub fn get_pool_id(
+			asset1: MultiAssetId<T::AssetId>,
+			asset2: MultiAssetId<T::AssetId>,
+		) -> PoolIdOf<T> {
+			if asset1 <= asset2 {
+				(asset1, asset2)
+			} else {
+				(asset2, asset1)
+			}
```
[the remark this answers is not in the document]
Comment (jsidorenko) [reply frame/dex/src/lib.rs:684]: I added a manual implementation of the Ord/PartialOrd trait for NativeOrAssetId enum

File: frame/dex/src/lib.rs [hunk 11]
```diff
@@ -0,0 +1,849 @@
...
+			let amount2 = Self::mul_div(&lp_token_burn, &reserve2, &total_supply)?;
+
+			ensure!(
+				!amount1.is_zero() && amount1 >= amount1_min_receive,
+				Error::<T>::InsufficientAmountParam1
+			);
+			ensure!(
+				!amount2.is_zero() && amount2 >= amount2_min_receive,
+				Error::<T>::InsufficientAmountParam2
+			);
```
[the remark this answers is not in the document]
Comment (jsidorenko) [reply frame/dex/src/lib.rs]: updated

File: bin/node/runtime/src/lib.rs [hunk 12]
```diff
@@ -1486,6 +1488,52 @@ impl pallet_assets::Config for Runtime {
...
+	type BenchmarkHelper = ();
+}
+
+parameter_types! {
+	pub const DexPalletId: PalletId = PalletId(*b"py/dexer");
+	pub AllowMultiAssetPools: bool = true;
+}
+
+impl pallet_dex::Config for Runtime {
+	type Fee = ConstU64<3>;
```
[the remark this answers is not in the document]
Comment (jsidorenko) [reply bin/node/runtime/src/lib.rs]: done

**Comment by jsidorenko** [comment]
bot bench $ pallet dev pallet_asset_conversion

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2615754

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2615740

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2616232

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2616218

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2616278

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2616292

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2702528

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2702527

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2704330

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2707748

File: frame/asset-conversion/src/lib.rs [hunk 13]
```diff
@@ -0,0 +1,1043 @@
...
+		type Currency: InspectFungible<Self::AccountId, Balance = Self::Balance>
+			+ MutateFungible<Self::AccountId>;
+
+		/// The `Currency::Balance` type.
+		type Balance: Balance;
+
+		/// The type used to describe the amount of fractions converted into assets.
+		type AssetBalance: Balance;
+
+		type HigherPrecisionBalance: IntegerSquareRoot
```
Comment (jsidorenko) [inline frame/asset-conversion/src/lib.rs:111]: Missed docs

File: frame/asset-conversion/src/lib.rs [hunk 14]
```diff
@@ -0,0 +1,1043 @@
...
+		/// Weight information for extrinsics in this pallet.
+		type WeightInfo: WeightInfo;
+
+		/// The benchmarks need a way to create asset ids from u32s.
+		#[cfg(feature = "runtime-benchmarks")]
+		type BenchmarkHelper: BenchmarkHelper<Self::AssetId>;
+	}
+
+	#[pallet::storage]
+	pub type Pools<T: Config> =
```
Comment (jsidorenko) [inline frame/asset-conversion/src/lib.rs:179]: Missed docs

File: frame/asset-conversion/src/lib.rs [hunk 15]
```diff
@@ -0,0 +1,1043 @@
...
+		},
+		Transfer {
+			from: T::AccountId,
+			to: T::AccountId,
+			asset: T::MultiAssetId,
+			amount: AssetBalanceOf<T>,
+		},
+	}
+
+	// Your Pallet's error messages.
```
Comment (jsidorenko) [inline frame/asset-conversion/src/lib.rs:229]:
```suggestion
```

File: frame/asset-conversion/src/lib.rs [hunk 16]
```diff
@@ -0,0 +1,1043 @@
...
+	/// Stores the `PoolAssetId` that is going to be used for the next lp token.
+	/// This gets incremented whenever a new lp pool is created.
+	#[pallet::storage]
+	pub type NextPoolAssetId<T: Config> = StorageValue<_, T::PoolAssetId, OptionQuery>;
+
+	// Pallet's events.
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		PoolCreated {
```
Comment (jsidorenko) [inline frame/asset-conversion/src/lib.rs:191]: We need to put doc for each event

File: frame/asset-conversion/src/lib.rs [hunk 17]
```diff
@@ -0,0 +1,1043 @@
...
+		/// The provided path must consists of 2 assets at least.
+		InvalidPath,
+		/// It was not possible to calculate path data.
+		PathError,
+	}
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<T::BlockNumber> for Pallet<T>
+	// where
+	// T::AssetId: From<u32>,
```
Comment (jsidorenko) [inline frame/asset-conversion/src/lib.rs:283]:
```suggestion
```

File: frame/asset-conversion/src/lib.rs [hunk 18]
```diff
@@ -0,0 +1,1043 @@
...
+		}
+
+		/// The account ID of the pool.
+		///
+		/// This actually does computation. If you need to keep using it, then make sure you cache
+		/// the value and only call this once.
+		pub fn get_pool_account(pool_id: PoolIdOf<T>) -> T::AccountId {
+			T::PalletId::get().into_sub_account_truncating(pool_id)
+			// let sub = sp_io::hashing::blake2_256(&Encode::encode(&pool_id)[..]);
+			// T::PalletId::get().into_sub_account_truncating(sub)
```
Comment (jsidorenko) [inline frame/asset-conversion/src/lib.rs:700]:
```suggestion
```

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2712552

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2712553

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2712902

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2712903

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2712969

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2712955

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2713834

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2713820

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2714993

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2714979

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2726176

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2726178

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2726177

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2726231

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2726230

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2726337

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2726545

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2726977

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2726963

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2727820

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2729864

File: frame/dex/src/lib.rs [hunk 19]
```diff
@@ -0,0 +1,849 @@
...
+//!  - swap 2 assets if there is a pool created
+//!  - query for an exchange price via a new runtime call endpoint
+//!
+//! Here is an example `state_call` that asks for a quote of a pool of native versus asset 1:
+//!
+//! ```text
+//! curl -sS -H "Content-Type: application/json" -d \
+//! '{"id":1, "jsonrpc":"2.0", "method": "state_call", "params": ["DexApi_quote_price", "0x0101000000000000000000000011"]}' \
+//! http://localhost:9933/
+//! ```
```
[the remark this answers is not in the document]
Comment (gilescope) [reply frame/dex/src/lib.rs]: all items are now documented.

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable-int
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2740688

**Comment by jsidorenko** [comment]
bot rebase

File: frame/asset-conversion/src/lib.rs [hunk 20]
```diff
@@ -0,0 +1,1187 @@
...
+		/// Provided assets are equal.
+		EqualAssets,
+		/// Pool already exists.
+		PoolExists,
+		/// Desired amount can't be zero.
+		WrongDesiredAmount,
+		/// Provided amount should be greater or equal to the existential deposit/asset's minimal
+		/// amount.
+		AmountLessThanMinimal,
+		/// Reserve needs to always be greater or equal to the existential deposit/asset's minimal
```
Comment (sam0x17) [inline frame/asset-conversion/src/lib.rs:307]:
```suggestion
		/// Reserve needs to always be greater than or equal to the existential deposit/asset's minimal
```

File: frame/asset-conversion/src/lib.rs [hunk 21]
```diff
@@ -0,0 +1,1187 @@
...
+
+	#[pallet::error]
+	pub enum Error<T> {
+		/// Provided assets are equal.
+		EqualAssets,
+		/// Pool already exists.
+		PoolExists,
+		/// Desired amount can't be zero.
+		WrongDesiredAmount,
+		/// Provided amount should be greater or equal to the existential deposit/asset's minimal
```
Comment (sam0x17) [inline frame/asset-conversion/src/lib.rs:304]:
```suggestion
		/// Provided amount should be greater than or equal to the existential deposit/asset's minimal
```

File: frame/asset-conversion/src/lib.rs [hunk 22]
```diff
@@ -0,0 +1,1187 @@
...
+		AssetTwoDepositDidNotMeetMinimum,
+		/// The minimal amount requirement for the first token in the pair wasn't met.
+		AssetOneWithdrawalDidNotMeetMinimum,
+		/// The minimal amount requirement for the second token in the pair wasn't met.
+		AssetTwoWithdrawalDidNotMeetMinimum,
+		/// Optimal calculated amount is less than desired.
+		OptimalAmountLessThanDesired,
+		/// Insufficient liquidity minted.
+		InsufficientLiquidityMinted,
+		/// Asked liquidity can't be zero.
```
Comment (sam0x17) [inline frame/asset-conversion/src/lib.rs:328]:
```suggestion
		/// Requested liquidity can't be zero.
```

File: frame/asset-conversion/src/lib.rs [hunk 23]
```diff
@@ -0,0 +1,1187 @@
...
+
+			Self::deposit_event(Event::PoolCreated { creator: sender, pool_id, lp_token });
+
+			Ok(())
+		}
+
+		/// Provide liquidity into the pool of `asset1` and `asset2`.
+		/// NOTE: an optimal amount of asset1 and asset2 will be calculated and
+		/// might be different to provided `amount1_desired`/`amount2_desired`
+		/// thus it's needed to provide the min amount you're happy to provide.
```
Comment (sam0x17) [inline frame/asset-conversion/src/lib.rs:438]:
```suggestion
		/// thus you should provide the min amount you're happy to provide.
```

File: frame/asset-conversion/src/lib.rs [hunk 24]
```diff
@@ -0,0 +1,1187 @@
...
+				amount2,
+				lp_token: pool.lp_token,
+				lp_token_burned: lp_token_burn,
+			});
+
+			Ok(())
+		}
+
+		/// Swap the exact amount of `asset1` into `asset2`.
+		/// `amount_out_min` param allows to specify the min amount of the `asset2`
```
Comment (sam0x17) [inline frame/asset-conversion/src/lib.rs:632]:
```suggestion
		/// `amount_out_min` param allows you to specify the min amount of the `asset2`
```

**Review by sam0x17 — COMMENTED** [review]
Looks good at a high level. Will approve once others have weighed in. Found a few doc comment/grammar tweaks :)

**Comment by jsidorenko** [comment]
bot rebase

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2872687

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2872919

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2883967

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate-macos
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2890073

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate-macos
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2890842

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate-macos
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2893706

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2893689

**Comment by jsidorenko** [comment]
let's allow auditors to set the "audited" label :)

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2894764

**Review by jsidorenko — APPROVED** [review]

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: check-dependent-cumulus
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2897564

**Review by joepetrowski — APPROVED** [review]

**Comment by gilescope** [comment]
bot merge
